---
title: "DNA methylation analysis and integration using a real dataset"
author: "Edoardo “Dado” Marcora"
date: "September 12, 2012"
output: html_document
---

Jaffe AE, Murakami P, Lee H, Leek JT, Fallin MD, Feinberg AP, Irizarry RA (2012) Bump hunting to identify differentially methylated regions in epigenetic epidemiology studies. International Journal of Epidemiology 41(1):200-9.

This is not a toy example but rather uses a dataset similar to those being analyzed in academic research labs, which is available at: https://github.com/genomicsclass/tcgaMethylationSubset/archive/master.zip

```{r}
library(rafalib)
library(minfi) # Bioc
library(IlluminaHumanMethylation450kmanifest) # Bioc
library(IlluminaHumanMethylation450kanno.ilmn12.hg19) # Bioc
library(doParallel) # CRAN

tryCatch(load("grset.rda"), error = function(c) {
  grset <- getGenomicRatioSetFromGEO("GSE32148")
  save(grset, file = "grset.rda")
})
```

[As seen previously](index.html), we use a functional model in which each methylation measurement is treated in the context of other measurements that are nearby in the genome, i.e., as a function (loess or $l$) of genomic location to take into account the regional context/other measurements in the region:

![Functional CpG model](./functional_cpg_model.png)

This is based on the hypothesis that epigenetic marks can affect gene expression only when changing coordinately across a region.

The model above can be used to find differentially methylated regions (DMRs) across, e.g., normal and cancer samples (modeled using the dummy variable $X_i$ coded as 0 and 1), or even a continuos trait, e.g., age (modeled using the continuos variable $X_i$).

First, we calculates the effect size estimates for each CpG by applying the simple or functional model. The simple model (`bumphunter` function with smooth argument = FALSE) returns the effect size estimates for each CpG irrespective of nearby CpGs.

![](bumphunter_1.png)

From this we still need to derive the effect size estimates of regions of interest, e.g., the area (length * average height, or the length and average height separately) of "bumps" below/above a certain threshold, from which we can calculate significance and thus DMRs using, e.g., permutations to obtain a null distribution of bumps!

![](bumphunter_2.png)

![](bumphunter_3.png)

First, we load the data and compute the methylation values using the Illumina protocol.

```{r}
getwd()
path <- "tcgaMethylationSubset"
targets <- read.delim(file.path(path, "targets.txt"), as.is = TRUE)
table(targets$Tissue, targets$Status)

# How many samples are represented in this table?
nrow(targets)

# How many samples are normal colon samples?
sum(targets$Tissue=="colon" & targets$Status=="normal")

# restrict analysis to a comparison of normal "colon" vs "breast" samples for demonstration purposes
idx <- which(targets$Status=="normal" & targets$Tissue %in% c("colon", "breast"))
targets <- targets[idx, ]

# load the idat files
dat <- read.450k.exp(base = path, targets = targets, verbose = TRUE)

# preprocess the data using the Illumina protocol
dat <- preprocessIllumina(dat)

# assign genomic locations to each CpG
dat <- mapToGenome(dat)

# compute methylation/Beta values from U and M values
dat <- ratioConvert(dat, type="Illumina")

class(dat)
```

Next, we perform exploratory data analysis to spot outlier samples.

```{r}
# inspect the distribution of the methylation/Beta values across all the samples
mypar(1,1)
y = getBeta(dat)
shist(y)

# color samples by tissue
cols = as.factor(pData(dat)$Tissue)

# create an MDS plot to search for outlier samples
mds = cmdscale(dist(t(y)))
plot(mds, col = cols)
```

No sample stands out as an outlier! All samples have similar distributions of methylation/Beta values and the first PC splits the data by tissue as expected.

First, we analyze the data using the `limma` package to identify CpGs that differ significantly in their methylation status in "colon" vs "breast" samples.

```{r}
library(limma)

# create design matrix
tissue = as.factor(pData(dat)$Tissue)
X = model.matrix(~ tissue)

# extract methylation values
y = getBeta(dat)

# obtain effect sizes and pvals of CpGs with limma
fit = lmFit(y, X)
eb <- ebayes(fit)
library(qvalue)
qvals = qvalue(eb$p.value[, 2])$qvalue

# find the genomic location and q-value of the CpG with the largest effect size
tophit_idx = which.max(abs(fit$coef[, 2]))
seqnames(dat[tophit_idx])
start(dat[tophit_idx])
qvals[tophit_idx]

# find all the CpGs within 5Kb of the location of this CpG
thiscpg <- granges(dat[tophit_idx])
dat_idx <- which(granges(dat) %over% (thiscpg + 5000))

# positions of these CpGs
pos <- start(dat)[dat_idx]

# set xlim to include all of these CpGs
xlim <- range(c(pos))

# create a plot showing the methylation values for all samples for these CpGs. Use color to distinguish "breast" from "colon" samples.

# plot the beta/methylation values for these CpGs
mypar(3, 1)
matplot(pos, y[dat_idx, , drop = FALSE], col = cols, pch = 20, xlim = xlim, xlab = "genomic location", ylab = "methylation")

# Plot the estimated effect sizes and the -log10 of the q-values for these CpGs in two separate plots
plot(pos, fit$coef[dat_idx, 2], pch = 20, xlim = xlim, ylim = c(-1,1), xlab = "genomic location", ylab = "slope/effect size")
abline(h=c(-0.1,0,0.1))

plot(pos, -log10(qvals[dat_idx]), pch = 20, xlim = xlim, xlab = "genomic location", ylab = "-log10(q-value)")

# Repeat the above for the top 10 CpGs ranked by absolute value of effect size
mypar(3,1)
o = order(abs(fit$coef[,2]), decreasing = TRUE)[1:10]
for(i in o){
  tophit_idx = i
  thiscpg <- granges(dat[tophit_idx])
  dat_idx <- which(granges(dat) %over% (thiscpg + 5000))
  pos <- start(dat)[dat_idx]
  xlim <- range(c(pos))

  matplot(pos, y[dat_idx, , drop = FALSE], col = cols, pch = 20, xlim = xlim, xlab = "genomic location", ylab = "methylation")

  plot(pos, fit$coef[dat_idx, 2], pch = 20, xlim = xlim, ylim = c(-1,1), xlab = "genomic location", ylab = "slope/effect size")
  abline(h=c(-0.1,0,0.1))

  plot(pos, -log10(qvals[dat_idx]), pch = 20, xlim = xlim, xlab = "genomic location", ylab = "-log10(q-value)")
}
```

Second, we analyze the data using the `bumphunter` package to identify regions (DMRs) that differ significantly in their methylation status in "colon" vs "breast" samples.

```{r}
# parallelize over multiple cores for faster processing of bumphunter permutations
library(doParallel)
ncores = detectCores()
registerDoParallel(cores = ncores)

# setup model (and restrict analysis to chr15 for demonstration purposes)
tissue <- pData(dat)$Tissue
X <- model.matrix(~ tissue)
idx <- which(seqnames(dat)=="chr15")
dat2 <- dat[idx, ]
set.seed(1)

# obtain effect sizes and pvals of DMRs with bumphunter (here using the functional model, i.e., the default smooth argument = TRUE, use FALSE for the simple model and to improve performance)
res <- bumphunter(dat2, X, cutoff = 0.1, B = 1000) # B is the number of permutations to perform in order to obtain a null distribution for hypothesis testing

class(res)
head(res$table)

# According to these results, how many regions achieve an FWER lower than 0.05, i.e. are DMRs?
sum(res$table$fwer < 0.05)
```

Please note that, when calculating p-values based on permutations, a p-value of 0 (which is non-sensical) obtained by performing 1000 permutations instead it really means p-value < 1/1000 (p-value < 1e-3)!

Next, we inspect the relationship between the CpGs returned by `limma` that achieve q-values smaller than 0.05 and an effect size larger than 0.5 (in absolute value), and of the regions returned by `bumphunter` that contains 3 or more CpGs:

```{r}
index = which(qvals < 0.05 & abs(fit$coef[, 2]) > 0.5 & seqnames(dat)=="chr15")

tab = res$table[res$table$L >= 3, ]
tab = makeGRangesFromDataFrame(tab)

# What proportion of the CpGs indexed by index overlap with regions inside tab?
mean(countOverlaps(granges(dat[index]), tab))

# or more simply

mean(granges(dat[index]) %over% tab)
```

Next, we inspect the relationship between the differentially methylated regions (DRMs) in "colon" vs "breast" samples, and the CpG islands annotated in the reference human genome:

```{r}
# retrieve the genomic locations of CpG islands in the reference human genome using the AnnotationHub
library(AnnotationHub)
cgi <- AnnotationHub()[["AH5086"]]

# convert the bumphunter output to a GRanges object that contains all of the regions that are differentially methylated in "colon" vs "breast" samples with a FWER statistical significance threshold of .05
tab <- res$table[res$table$fwer <= .05,]
tab <- makeGRangesFromDataFrame(tab, keep.extra.columns = TRUE)

# calculate the distance of each DMR to the nearest CpG island and display a summary table
map <- distanceToNearest(tab, cgi)
d <- mcols(map)$dist
prop.table(table(cut(d, c(0, 1, 2000, 5000, Inf), include.lowest = TRUE, right = FALSE)))
```

Please note that there are more DMRs that are within 2Kb of a CpG island (aka CpG shore) as there are DMRs within a CpG island. In addition, there are some DMRs that are more than 2Kb from a CpG island (aka CpG shelve) and some that are more than 5Kb away from a CpG island (aka CpG open-sea)!

In conclusion, DMRs are more frequenly observed in CpG island shores than in CpG islands! This is not the case when inspecting the null hypothesis scenario:

```{r}
null <- granges(dat)
nulltab <- makeGRangesFromDataFrame(null, keep.extra.columns = TRUE)

nullmap <- distanceToNearest(nulltab, cgi)
nulld <- mcols(nullmap)$dist
prop.table(table(cut(nulld,c(0, 1, 2000, 5000, Inf), include.lowest=TRUE, right=FALSE)))

# or more simply

prop.table(table(getIslandStatus(dat)))
```


```{r}
# sort the DMRs by area
tab <- tab[order(-mcols(tab)$area)]

# add 3Kb to each side of a DMR
tab <- tab + 3000

# let's analyze DMR # 17 as an example
i <- 17
dat_idx <- which(granges(dat) %over% tab[i]) # CpGs overlapping with DMR # 17
cgi_idx <- which(cgi %over% tab[i]) # CpGs island overlapping with DMR # 17
thecgi <- cgi[cgi_idx]

# positions of CpGs overlapping with DMR # 17
pos <- start(dat)[dat_idx]

# set xlim to include both the CpG region and the CpG island entirely
xlim <- range(c(pos, start(thecgi), end(thecgi)))

# get the beta/methylation values for the CpGs overlapping with DMR # 17
beta <-  getBeta(dat)
y <- beta[dat_idx, ]

# plot the beta/methylation values for the CpGs overlapping with DMR # 17
mypar(1, 1)
matplot(pos, y, col = cols, pch = 20, xlim = xlim, ylim = c(0,1), xlab = "genomic location", ylab = "methylation")
# add an indication of the CpGs island overlapping with DMR # 17
apply(cbind(start(thecgi), end(thecgi)), 1, function(x) segments(x[1], 0, x[2], 0, lwd=4, col=3))

# plot the methylation difference of "colon" vs "breast" for the CpGs overlapping with DMR # 17
plot(pos, res$fitted[dat_idx], xlim = xlim, ylim = c(-0.4,0.4))
abline(h=0)
# add an indication of the CpGs island overlapping with DMR # 17
apply(cbind(start(thecgi), end(thecgi)), 1, function(x) segments(x[1], 0, x[2], 0, lwd=4, col=3))
```

Here we are going to study the relationship between gene expression and DNA methylation by integrating gene expression and DNA methylation high throughput data. We will compare colon and lung samples.

```{r}
targets = read.delim(file.path(path, "targets.txt"), as.is = TRUE)
index = which(targets$Status=="normal" & targets$Tissue %in% c("colon", "lung"))
targets = targets[index, ]
dat = read.450k.exp(base = path, targets = targets, verbose = TRUE)
dat = preprocessIllumina(dat)
dat = mapToGenome(dat)
dat = ratioConvert(dat, type = "Illumina")
tissue = pData(dat)$Tissue
X = model.matrix(~ tissue)
res = bumphunter(dat, X, cutoff = 0.25)

nrow(res$table)
head(res$table)

# What proportion of regions are just one CpG?
mean(res$table$L==1)
```

Now we are going to match these regions to genes. First, we load the related gene expression data. Note that this will load an object called tcgaLungColonExpLM which is the result (linear model fit) of a differential expression analysis using `limma` on the raw TCGA microarray data for lung vs colon samples:

```{r}
load(file.path(path, "tcgaLungColonExpLM.rda"))
class(tcgaLungColonExpLM)

# note that we saved the annotation of the gene expression data in the `annotation` object:

print(annotation)
```

We can obtain q-values using the `qvalue` package.

```{r}
library(limma)
library(qvalue)
eb = ebayes(tcgaLungColonExpLM)
qvals = qvalue(eb$p.value[, 2])$qvalue
```

Now we obtain the locations of these differentially expressed genes.

```{r}
# biocLite("hgu133plus2.db")
library(hgu133plus2.db)
map = select(hgu133plus2.db, keys = rownames(tcgaLungColonExpLM$coef), columns = c("ENTREZID"), keytype = "PROBEID")

library(Homo.sapiens)
Genes = genes(Homo.sapiens)
Genes = resize(Genes, 1) # to obtain the tss

index1 = match(as.character(mcols(Genes)$GENEID), map$ENTREZID)
index2 = match(map$PROBEID[index1], rownames(tcgaLungColonExpLM$coef))
M = tcgaLungColonExpLM$coef[index2, 2]
```

M is now in the same order as Genes. We can now find the closest gene to each DMR.

```{r}
tab = makeGRangesFromDataFrame(res$table, keep.extra.columns = TRUE)
map2 = distanceToNearest(tab, Genes)
```

Now we can make plots comparing the methylation differences to the gene expression differences. We consider DMRs of different sizes:

```{r}
index1 = subjectHits(map2)
dist = mcols(map2)$dist

mypar(2,2)
for (i in c(0, 1, 2, 3)) {
  keep = dist < 10000 & tab$L > i
  plot(tab$value[keep], M[index1][keep], main = paste("cor=", signif(cor(tab$value[keep], M[index1][keep], use = "complete"), 2)))
}
```

## Accounting for cell composition when analyzing whole-tissue extracts

Tissues, e.g., blood, are a mixture of many cell types, each with distinct DNA methylation patterns/signatures.

To examine the importance of accounting for cellular composition in DNA methylation analysis, we are going to download a GEO dataset used in an analysis of whole blood data.

The minfi package has a function to read data directly from GEO.

```{r}
library(minfi)
grset = getGenomicRatioSetFromGEO("GSE32148")
class(grset)
```

With a bit of wrangling we can extract the age as a number of the individuals sampled here:

```{r}
colnames(pData(grset))

pData(grset)[1, ]
pData(grset)[1, "characteristics_ch1.1"] # age column

age = pData(grset)$characteristics_ch1.1
age = as.character(age)
age[grep("N/A", age)] = NA
age = as.numeric(gsub("age \\(y\\):\ ", "", age))
```

This experiment was performed to find DMRs between individuals with Crohn's disease and controls. We can extract this information like this:

```{r}
group = rep("normal", nrow(pData(grset)))
group[grepl("ulcerative", pData(grset)[, 1])]="ulcerative"
group[grepl("Crohn", pData(grset)[, 1])]="crohn"
group = factor(group, levels = c("normal", "ulcerative", "crohn"))
```

Before EDA, we remove CpGs with NA calls as well as the sex chromosomes:

```{r}
keep = which(rowSums(is.na(getBeta(grset)))==0 & !seqnames(grset) %in% c("chrX", "chrY"))
grset2 = grset[keep, ]
```

Use a multi-dimensional scaling (MDS) plot on the methylation measurements in grset2. Use color to denote individuals that are older than 40 and different symbols to denote the different conditions: normal, ulcerative, crohn.

```{r}
# inspect the distribution of the methylation/Beta values across all the samples
mypar(1,1)
y = getBeta(grset2)
shist(y)

# create an MDS plot to search for outlier samples
mds = cmdscale(dist(t(y)))
col=ifelse(age > 40, 1, 2)
plot(mds, pch = as.numeric(group), col = col)
legend("bottomleft", levels(group), pch = seq_along(group))
```

The `minfi` package provides the function `estimateCellCounts` to estimate the cell composition of a tissue. Currently, this function is recommended to be used starting with raw data, not processed data from GEO. This is a topic of current research and might change in the near future.

## Multi-resolution analysis of DNA methylation

Here, we search for differentially methylated blocks:

```{r}
library(minfi)
targets = read.delim(file.path(path, "targets.txt"), as.is = TRUE)
index = which(targets$Tissue=="colon")
targets = targets[index, ]
dat = read.450k.exp(base=path, targets = targets, verbose = TRUE)
dat = preprocessIllumina(dat)
dat = mapToGenome(dat)
```

Now, we collapse the data:

```{r}
cdat = cpgCollapse(dat)
class(cdat)
?cpgCollapse
class(cdat$object)
class(cdat$blockInfo)

# The original data include nrow(dat) CpGs
nrow(dat)

# The collapsed object include nrow(cdat$object) regions
nrow(cdat$object)
```

We can see the type of regions that are represented in this collapsed object:

```{r}
head(granges(cdat$object))

# What proportion of the regions are OpenSea regions?
mean(mcols(granges(cdat$object))$type=="OpenSea")

prop.table(table(mcols( granges(cdat$obj))$type))
```

Now we use the `blockFinder` function to find differentially methylated regions between cancer and normal colon samples:

```{r}
status = factor(pData(cdat$obj)$Status, level = c("normal", "cancer"))
X = model.matrix(~ status)
res = blockFinder(cdat$obj, X, cutoff = 0.05)
```

The `blockFinder` function calls `bumphunter` and returns a similar object:

```{r}
head(res$table)

# What proportion of the blocks reported in res$table are hypomethyated (lower methylation in cancer versus normal)?

mean(res$table$value < 0)
```

```{r}
tab = makeGRangesFromDataFrame(res$table)
index = granges(cdat$obj) %over% (tab[1] + 10000)
pos = start(cdat$obj)[index]
col = as.numeric(status)
matplot(pos,getBeta(cdat$obj)[index,],col = col,pch = 1,cex = 0.5)
plot(pos,res$fitted[index])
```


## Reduced Representation Bisulfite Sequencing

Reduced Representation Bisulfite Sequencing (RRBS) is an experimental technique widely used to manipulate the regions of the genome we measure. An enzyme is used to cut DNA at CCGG and the general idea is to filter out small or large molecules once DNA is cut. We can use Bioconductor tools to predict the size of these regions.

Load the genome package and create an object with the sequence for chr22:

```{r}
library("BSgenome.Hsapiens.UCSC.hg19")
chr22 = Hsapiens[["chr22"]]
```

Now use the `matchPattern` function to find all the locations in which CCGG occurs on chr22.

```{r}
res = matchPattern("CCGG", chr22)

# How many CCGG do we find on chr22?
length(res)

# plot a histogram of the DNA fragment sizes after we cut chr22 at CCGG
size = diff(start(res))
hist(size)
hist(log10(size))
```

A typical size to filter are DNA regions between 40 and 220 basepairs.

```{r}
# What proportion of the fragments created for chr22 are between 40 and 220 basepairs?
mean(size >= 40 & size <= 220)
```

If we sequence all of chromosome 22 we need to sequence 51,304,566bp However, if instead we keep only fragments of size between 40 and 220 basepairs, how much sequence would would be left?

```{r}
sum(size[size >= 40 & size <= 220])
```

Now, we use a set of paired tumor and normal colon whole-genome bisulfite sequencing (WGBS) data. You can download the data we will use from [here](https://github.com/genomicsclass/colonCancerWGBS). More information about these data is available on GEO [GSE46644](http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE46644). Here, we use the .cov files from Bismark as input to the `bsseq` package to create a `bsseq` object. The data come from this paper.

The step-by-step instruction on how to covert the raw fastq files into the files we work on here are available [here](https://github.com/genomicsclass/colonCancerWGBS/blob/master/scripts/createObject.Rmd).

Let's start by the reading in the sample information.

```{r}
path = "colonCancerWGBS"
targets = read.delim(file.path(path, "targets.txt"), as.is = TRUE)
head(targets)
```

Now you will need the `bsseq` package to read in the sequencing data.

We load the methylation calls from our alignments into R. Once the data are loaded into R, we can use this package for further downstream analyses such as finding differentially methylated regions between our paired tumor and normal colon cancer samples. This package assumes the following data has been extracted from the alignments:

1) genomic positions (chromosome and location) of CpGs
2) M (Methylation) values, describing the number of read supporting methylation covering a single CpG
3) Cov (Coverage) values, describing the number of read supporting methylation covering a single CpG

For illustrative purposes, we only consider the methylation loci in chromosome 22 from the .cov file.

```{r}
# biocLite("bsseq")
library("bsseq")
cov.files = list.files(path = path, pattern = "*chr22.cov", full.names = TRUE) # coverage files
colonCancerWGBS = read.bismark(files = cov.files, rmZeroCov = TRUE, sampleNames = as.character(targets$Run)) # you can ignore the warning message

# add sample information to object
colData(colonCancerWGBS) = DataFrame(targets) 
```  

To view the `bsseq` object and the phenotypic information about each sample:
```{r}
colonCancerWGBS

# phenotypic information
pData(colonCancerWGBS)

# granges object
granges(colonCancerWGBS)
```

Now we can extract the coverage and the number of reads with evidence of methylation:

```{r}
cov = getCoverage(colonCancerWGBS, type = "Cov")
head(cov)

m = getCoverage(colonCancerWGBS, type = "M")
head(m)

# What proportion of the reported CpGs have some coverage in all sample?
index = apply(cov > 0, 1, all)
mean(index)

# Compute the total coverage (across all samples) for each CpG and plot it against genomic location
tot = rowSums(cov)
hist(tot)
loc = start(colonCancerWGBS)
for (i in 1:11) {
  index = 1:100000 + 100000 * i # very ad-hoc
  plot(loc[index], tot[index], cex = .5, ylim = c(0,300))
}
```

Note that we can get coverage (`cov`) and the number of reads including evidence for methylation (`m`) as shown above and assuming cov > 0, `m/cov` gives us an estimate of methylation status at each CpG and `1/sqrt(cov)` gives us a quantity proportional to the standard error of this estimate.

```{r}
gr = GRanges(seqnames = "22", ranges = IRanges(start = 43793678, end = 45022550))
index = granges(colonCancerWGBS) %over% gr
i = 1
index2 = which(index & cov[, i] >= 5 & cov[, i] <= 50)
x = start(colonCancerWGBS)[index2]
y = m[index2, i] / cov[index2, i]
w = sqrt(cov[index2, i]) / 7
plot(x, y, cex = w)
```